Backport of:

From 54f90717c1d4e5846d8f99c212048f26796beef4 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Wed, 8 Feb 2017 13:38:04 -0500
Subject: [PATCH] Fix an out of bound error in sun file handling

bug: https://github.com/ImageMagick/ImageMagick/issues/375
bug: https://github.com/ImageMagick/ImageMagick/issues/376
bug-debian: https://bugs.debian.org/856879
origin: https://github.com/ImageMagick/ImageMagick/commit/3007531bfd326c5c1e29cd41d2cd80c166de8528

(cherry picked from commit 3007531bfd326c5c1e29cd41d2cd80c166de8528)
---
 coders/sun.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

Index: imagemagick-6.6.9.7/coders/sun.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/sun.c	2017-03-14 10:30:27.888471826 -0400
+++ imagemagick-6.6.9.7/coders/sun.c	2017-03-14 11:01:00.085389263 -0400
@@ -181,6 +181,21 @@
   return(MagickTrue);
 }
 
+static MagickBooleanType HeapOverflowSanityCheck(const size_t count,
+  const size_t quantum)
+{
+  size_t
+    size;
+
+  size=count*quantum;
+  if ((count == 0) || (quantum != (size/count)))
+    {
+      errno=ENOMEM;
+      return(MagickTrue);
+    }
+  return(MagickFalse);
+}
+
 /*
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 %                                                                             %
@@ -257,7 +272,8 @@
   size_t
     bytes_per_line,
     extent,
-    height;
+    height,
+    pixels_length;
 
   ssize_t
     count,
@@ -429,8 +445,13 @@
     if ((bytes_per_line >> 1) != (sun_info.width*sun_info.depth+15))
       ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
     bytes_per_line>>=4;
-    sun_pixels=(unsigned char *) AcquireQuantumMemory(height,
-      (MagickMax(image->columns,bytes_per_line)+1)*sizeof(*sun_pixels));
+    if (HeapOverflowSanityCheck(height,MagickMax(image->columns,bytes_per_line)+1) != MagickFalse)
+      {
+        sun_data=(unsigned char *) RelinquishMagickMemory(sun_data);
+        ThrowReaderException(ResourceLimitError,"ImproperImageHeader");
+      }
+    pixels_length=height*(MagickMax(image->columns,bytes_per_line)+1);
+    sun_pixels=(unsigned char *) AcquireQuantumMemory(pixels_length,sizeof(*sun_pixels));
     if (sun_pixels == (unsigned char *) NULL)
       ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
     ResetMagickMemory(sun_pixels,0,height*(MagickMax(image->columns,
